src/rust/engine/process_execution/src/docker.rs

```diff
@@ -681,14 +697,86 @@ impl<'a> ContainerCache<'a> {
     Ok(container.id)
   }
 
-  /// Return the container ID of a container running `image` for use as a place to invoke
-  /// build actions as executions within the cached container.
-  pub async fn container_id_for_image(
+  /// Creates a volume for named caches for the given image name. In production usage, the image
+  /// name will have been expanded to include its SHA256 fingerprint, so the named cache will be
+  /// dedicated to a particular image version. That is conservative, so we might consider making
+  /// it configurable whether the image version is attached in future releases.
+  async fn maybe_make_named_cache_volume(
+    docker: &Docker,
+    image: &str,
+  ) -> Result<String, DockerError> {
+    let image_hash = Digest::of_bytes(image.as_bytes())
```

Comment:
If I understand this correctly, we're taking an image name (which somehow we're promised will include the image's SHA256), hashing that, and using it as an ID. 

Why not just use the Docker image ID (which is its SHA256) directly? It would be easier to reason about during debugging, for one thing. If I see a volume called `pants-named-caches-c0668a6d2dae` I know it is associated with that image, and not with some image that has an expanded name that hashes to that (how would I then find that image?)

Furthermore, the same image may have multiple names, the ID is the canonical thing. We don't need separate named caches for each different way of referencing the same image.

I don't know how the general docker environment code identifies images, but it seems like it should be by ID, not by name, of which, again, their can be many.

Reply:
I don't see how conservatism plays into this. The docker image ID is guaranteed to uniquely identify the image. Instead of tacking it on to the end of the image name (again, one of potentially many for the image) and then hashing that string, why not just use it directly?

Reply:
> The docker image ID is guaranteed to uniquely identify the image. Instead of tacking it on to the end of the image name (again, one of potentially many for the image) and then hashing that string, why not just use it directly?

Because it contains illegal characters, as mentioned above. It would need to be parsed.

Reply:
The ID is just a hex string. E.g., 

```
$ docker images
REPOSITORY       TAG                 IMAGE ID       CREATED         SIZE
python           3.9                 c0668a6d2dae   2 months ago    865MB
python           latest              16c345df4509   2 months ago    863MB
busybox          1                   2d0d8216f525   3 months ago    1.46MB
test-image       1.0                 56a43808849a   3 months ago    862MB
python           3.10.7-alpine3.16   40d35c171a81   4 months ago    47.9MB
busybox          1.34.1              4a49b68e7c0a   4 months ago    1.38MB
...
```

Reply:
I guess what you get in `   image: &str,` is something like `test-image:1.0@sha256: 56a43808849a...` and that is what needs to be parsed to get the image id.

Reply:
Background: When I originally wrote this Docker executor code, I found that parsing the Docker image name / tag is harder than it looks to do it right. See https://github.com/pantsbuild/pants/issues/17104 and the link to the podman code for how complex they made it. :(

Reply:
Can't `create_image` be used to isolate a pull to a separate step or is that not idempotent or else slow or...? That hands you back the image hash.

Reply:
> Can't `create_image` be used to isolate a pull to a separate step or is that not idempotent or else slow or...? That hands you back the image hash.

That is what @tdyas added in #17062: https://github.com/pantsbuild/pants/blob/8935264eea79ffd8cbd40b7067e4f2216a1aae28/src/rust/engine/src/intrinsics.rs#L702-L722 ... but AFAICT gives you back the string format that @tdyas / @kaos mentioned above.